src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -245,6 +245,62 @@ bool AbstractLQPNode::is_column_nullable(const ColumnID column_id) const {
   return left_input()->is_column_nullable(column_id);
 }
 
+std::vector<FunctionalDependency> AbstractLQPNode::functional_dependencies() const {
+  // Gather input FDs
+  auto fds_left = std::vector<FunctionalDependency>();
+  auto fds_right = std::vector<FunctionalDependency>();
+  if (left_input()) {
+    fds_left = left_input()->functional_dependencies();
+  }
+  if (right_input()) {
+    fds_right = right_input()->functional_dependencies();
+  }
+
+  auto fds_in = std::vector<FunctionalDependency>();
+  if (fds_right.empty()) {
+    fds_in = fds_left;
+  } else {
+    // Remove duplicate FDs that might result from e.g. self-joins
+    for (const auto& fd_right : fds_right) {
+      bool duplicate = std::any_of(fds_left.begin(), fds_left.end(), [&fd_right](const auto& fd_left) {
+        return (fd_left.first.size() == fd_right.first.size() && fd_left.second.size() == fd_right.second.size() &&
```

Comment:
std::tie?

Reply:
```cpp
return std::tie(fd_left.first.size(), fd_left.second.size()) < std::tie(fd_right.first.size(), fd_right.second.size())
```
The code above does not compile. How would you do it?

Reply:
Why `<`?

Reply:
If it's not more compact or you think it's not better to understand, never mind. It think it's much easier to understand for such tasks, but that's just my taste.

Reply:
> Why `<`?

Oh, it should be `==`. The `<` operator was the one I found in various examples.

Reply:
I thought it makes sense to compare the `size()`-properties first before checking both unordered sets (left & right) for equality. Unnecessary (micro-)optimization?

Reply:
Yes. Also, it [happens anyway.](https://github.com/gcc-mirror/gcc/blob/1580b4793d88132747f587ac95ebc12d96fee5b9/libstdc%2B%2B-v3/include/bits/stl_tree.h#L1627)